mmv1/tools/doc-parser/index.mjs

```diff
@@ -0,0 +1,161 @@
+import {JSDOM} from "jsdom"
+import fetch from "node-fetch"
+import fs from "fs"
+import TurndownService from 'turndown'
+var turndownService = new TurndownService()
+
+
+const host = "https://cloud.google.com"
+const path = "/dlp/docs/reference/rest/v2/projects.deidentifyTemplates#DeidentifyTemplate.CryptoReplaceFfxFpeConfig"
+const subObject = path.split('#').pop().split(".").join("\\.")
+const responses = {}
+
+async function fetchUrlContents(path){
+  var url = host + path
+  var absPath = path.split("#")[0]
+  if( !responses[absPath]) {
+    responses[absPath] = await (await fetch(url)).text()
+  }
+  return responses[absPath]
+}
+
+function htmlToMarkDown(element) {
+  return turndownService.turndown(element.outerHTML)
+}
+
+function analyzeDescription(elements) {
+  elements = [... elements]
+  var description = elements.map( e => htmlToMarkDown(e)).join('\n\n').trim()
+  var required = false
+  if (description.indexOf("Required.") == 0 ){
+    description = description.substring('Required.'.length)
+    required = true
+  }
+  return {description, required}
+}
+
+function printLine(tabIndex, line){
+  return "  ".repeat(tabIndex) + line + '\n'
+}
+
+function printDescription(tabIndex, description){
+  return printLine(
+    tabIndex, "description: | " + "\n" +
+    "  ".repeat(tabIndex+1) + description.split("\n").map(s=>s.trim()).join("\n" + "  ".repeat(tabIndex+1)).trim()
+  )
+}
+
+function shouldBeFiltered(key) {
+  var oneof = ['oneof_source', 'oneof_alphabet', 'oneof_characters', "oneof_transformation", "oneof_type"]
+  return oneof.findIndex(v => key.indexOf(v) != -1) == -1
+}
+
+async function generateEnumValues(path, object, tabIndex) {
+  var text = await fetchUrlContents(path)
+  var {window} = await new JSDOM(text);
+  var fields
+  if (!object) {
+    fields = window.document.querySelectorAll(`tr[id^=ENUM_VALUES]`)
+  } else {
+    fields = window.document.querySelectorAll(`tr[id^=${object}\\.ENUM_VALUES]`)
+  }
+  fields = [... fields]
+  fields.filter(field => field.getAttribute('id').indexOf('UNSPECIFIED') == -1)
+  var out = await Promise.all(fields.map( async field => {
+    var row = field.querySelectorAll('td')
+    row = [... row]
+    return "  ".repeat(tabIndex+1) + `- :${row[0].textContent.trim()}  #${row[1].textContent.trim().replaceAll("\n","")}`
+  }))
+  return out.join("\n")
+}
+
+async function generateProperties(window, object, tabIndex){
+  var fields
+  if (!object) {
+    fields = window.document.querySelectorAll(`tr[id^=FIELDS]`)
+  } else {
+    fields = window.document.querySelectorAll(`tr[id^=${object}\\.FIELDS]`)
+  }
+  fields = [... fields]
+  fields = fields.filter( element => shouldBeFiltered(element.getAttribute("id")))
+  var out = await Promise.all(fields.map( async element => {
+    console.log(element.getAttribute("id"))
+    var row = [... element.querySelectorAll(`td`)]
+    var name = row[0].textContent
+    if (row[1] == undefined) {
+      console.log("hi scott")
```

Comment:
Should most of these console.log lines be removed?

Reply:
I removed the hi, scott log. The other ones are useful to see the serialization process in action and that things are happening. I could hide it behind a debug flag or just leave it? wdyt?